<a href="https://colab.research.google.com/github/ummeafruz/Machine-Learning-Sessional/blob/main/Regression(Task1).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# Experiment Name: Exploring different types of Regression`

Experiment No: 01

Couse No: ETE 456 

Course Name: Neural and Fuzzy Systems in Communications Sessional

Submitted By:

                  Name: Umme Afruz

                  Id: 1608011

                  Level: 4

                  Term: II


**Objective**: 
 1. *To get familiar with regression in machine learning.*
 2. *To learn about different types of regression algorithms.*
 3. *To apply various regression algorithms on the real world dataset.*


### **Task**

### Prediction of Insurance Costs:

In [ ]:
import warnings
warnings.filterwarnings("ignore")

### Import the libraries:
Pandas is an open-source python library that is used for data manipulation and analysis. It provides many functions and methods to speed up the data analysis process. NumPy stands for 'Numerical Python'. It is an open-source Python library used to perform various mathematical and scientific tasks. It contains multi-dimensional arrays and matrices, along with many high-level mathematical functions that operate on these arrays and matrices.

In [ ]:
import numpy as np        
import pandas as pd     
import matplotlib.pyplot as plt   

### Dataset:

In [ ]:
# Download the data
!wget -O insurance.csv https://www.dropbox.com/s/mwgqgjbmfw0xa5p/insurance.csv?dl=0

--2021-12-20 18:54:31--  https://www.dropbox.com/s/mwgqgjbmfw0xa5p/insurance.csv?dl=0
Resolving www.dropbox.com (www.dropbox.com)... 162.125.80.18, 2620:100:601b:18::a27d:812
Connecting to www.dropbox.com (www.dropbox.com)|162.125.80.18|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: /s/raw/mwgqgjbmfw0xa5p/insurance.csv [following]
--2021-12-20 18:54:32--  https://www.dropbox.com/s/raw/mwgqgjbmfw0xa5p/insurance.csv
Reusing existing connection to www.dropbox.com:443.
HTTP request sent, awaiting response... 302 Found
Location: https://ucddd6c542343ab482bda8b6b4fb.dl.dropboxusercontent.com/cd/0/inline/BcPq5pE7lucrAbJpKwaB-IBtrk6kdOLt-6X7sUtZG-1bvNCgh65juA60GMKIvhjNYza4SzKragpDhHRCj8Kbhz-2Ommqq5pjC7Mew-ooks5zidf4EOb71pBdgUwATmbDZdzZEBbqmCOv2yfvH76C68TQ/file# [following]
--2021-12-20 18:54:33--  https://ucddd6c542343ab482bda8b6b4fb.dl.dropboxusercontent.com/cd/0/inline/BcPq5pE7lucrAbJpKwaB-IBtrk6kdOLt-6X7sUtZG-1bvNCgh65juA60GMKIvhjNYza4SzKragpDhHRC

## Importing the dataset 

In [ ]:
dataset = pd.read_csv('insurance.csv')
dataset

age     sex     bmi  children smoker     region      charges
0      19  female  27.900         0    yes  southwest  16884.92400
1      18    male  33.770         1     no  southeast   1725.55230
2      28    male  33.000         3     no  southeast   4449.46200
3      33    male  22.705         0     no  northwest  21984.47061
4      32    male  28.880         0     no  northwest   3866.85520
...   ...     ...     ...       ...    ...        ...          ...
1333   50    male  30.970         3     no  northwest  10600.54830
1334   18  female  31.920         0     no  northeast   2205.98080
1335   18  female  36.850         0     no  southeast   1629.83350
1336   21  female  25.800         0     no  southwest   2007.94500
1337   61  female  29.070         0    yes  northwest  29141.36030

[1338 rows x 7 columns]


Now splitting the data into independent & dependent variable is necessary.Independent are those that we are going to used for prediction or are the features. Dependent are those that we have to predict.

In [ ]:
feature = dataset[['age','sex','children','bmi','smoker','region']]  # for independent
target = dataset[['charges']]   # for dependent

In [ ]:
feature

age     sex  children     bmi smoker     region
0      19  female         0  27.900    yes  southwest
1      18    male         1  33.770     no  southeast
2      28    male         3  33.000     no  southeast
3      33    male         0  22.705     no  northwest
4      32    male         0  28.880     no  northwest
...   ...     ...       ...     ...    ...        ...
1333   50    male         3  30.970     no  northwest
1334   18  female         0  31.920     no  northeast
1335   18  female         0  36.850     no  southeast
1336   21  female         0  25.800     no  southwest
1337   61  female         0  29.070    yes  northwest

[1338 rows x 6 columns]

In [ ]:
target

charges
0     16884.92400
1      1725.55230
2      4449.46200
3     21984.47061
4      3866.85520
...           ...
1333  10600.54830
1334   2205.98080
1335   1629.83350
1336   2007.94500
1337  29141.36030

[1338 rows x 1 columns]

In [ ]:
dataset.isnull().value_counts()

age    sex    bmi    children  smoker  region  charges
False  False  False  False     False   False   False      1338
dtype: int64

There is no missing data or NAN value in this dataset.

#Encoding Categorial data
Categorical variables are usually represented as ‘strings’ or ‘categories’ and are finite in number. Machine learning models require all input and output variables to be numeric. This means that if your data contains categorical data, you must encode it to numbers before you can fit and evaluate a model.

### One Hot Encoding:
One hot encoding is a process of converting categorical data variables so they can be provided to machine learning algorithms to improve predictions. With one-hot, we convert each categorical value into a new categorical column and assign a binary value of 1 or 0 to those columns.

In [ ]:
from sklearn.preprocessing import LabelEncoder,OneHotEncoder

In [ ]:
encoder=OneHotEncoder(sparse=False)
encoded_labels = pd.DataFrame (encoder.fit_transform(feature[['smoker','sex','region']]))

In [ ]:
encoded_labels.columns = encoder.get_feature_names(['smoker','sex','region'])
dataset= pd.concat([feature, encoded_labels ], axis=1)

In [ ]:
dataset

age     sex  ...  region_southeast  region_southwest
0      19  female  ...               0.0               1.0
1      18    male  ...               1.0               0.0
2      28    male  ...               1.0               0.0
3      33    male  ...               0.0               0.0
4      32    male  ...               0.0               0.0
...   ...     ...  ...               ...               ...
1333   50    male  ...               0.0               0.0
1334   18  female  ...               0.0               0.0
1335   18  female  ...               1.0               0.0
1336   21  female  ...               0.0               1.0
1337   61  female  ...               0.0               0.0

[1338 rows x 14 columns]

In [ ]:
dataset.columns

Index(['age', 'sex', 'children', 'bmi', 'smoker', 'region', 'smoker_no',
       'smoker_yes', 'sex_female', 'sex_male', 'region_northeast',
       'region_northwest', 'region_southeast', 'region_southwest'],
      dtype='object')

In [ ]:
new_features = dataset[['age','bmi','children','smoker_no', 'smoker_yes','sex_female', 'sex_male','region_northeast', 'region_northwest','region_southeast', 'region_southwest']]

### Splitting the Dataset:
The train-test split is a technique for evaluating the performance of a machine learning algorithm. It can be used for classification or regression problems and can be used for any supervised learning algorithm.

**Train Dataset:** Used to fit the machine learning model.

**Test Dataset:** Used to evaluate the fit machine learning model.


In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
"""Spliting the Dataset into Training Set and Test Set """

X_train,X_test,y_train,y_test=train_test_split(new_features,target,test_size = 0.25,random_state = 0)
# random_state = 0 is select to get the same result

In [ ]:
print(X_train.shape)
print(X_test.shape)

(1003, 11)
(335, 11)


In [ ]:
X_train

age    bmi  ...  region_southeast  region_southwest
1075   32  29.59  ...               1.0               0.0
131    61  22.04  ...               0.0               0.0
15     19  24.60  ...               0.0               1.0
1223   20  24.42  ...               1.0               0.0
1137   26  22.23  ...               0.0               0.0
...   ...    ...  ...               ...               ...
763    27  26.03  ...               0.0               0.0
835    42  35.97  ...               1.0               0.0
1216   40  25.08  ...               1.0               0.0
559    19  35.53  ...               0.0               0.0
684    33  18.50  ...               0.0               1.0

[1003 rows x 11 columns]

Different types of Regression Algorithms are given below:


1. Linear Regression (Univariate or Multivariate)

2. Support Vector Regression

3. Decision Tree Regression

4. Random Forest Regressrion

### Multiple linear regression
Multiple Linear Regression is one of the important regression algorithms which models the linear relationship between a single dependent continuous variable and more than one independent variable.

In [ ]:
# Fitting Multiple Linear Regression Model to Training Set
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()

In [ ]:
regressor.fit(X_train,y_train)

LinearRegression()

In [ ]:

# Predicting the Test Set Results
y_pred = regressor.predict(X_test)

In [ ]:
y_test

charges
578    9724.53000
610    8547.69130
569   45702.02235
1034  12950.07120
198    9644.25250
...           ...
574   13224.05705
1174   4433.91590
1327   9377.90470
817    3597.59600
1337  29141.36030

[335 rows x 1 columns]

In [ ]:
y_pred

array([[11121.10140924],
       [ 9369.08345223],
       [38349.2588068 ],
       [16331.93500645],
       [ 7041.22770615],
       [ 4103.24923177],
       [ 1734.03805961],
       [14478.86491464],
       [ 9168.77981235],
       [ 7715.94671029],
       [ 4625.04765547],
       [10398.53755318],
       [ 8756.48015868],
       [ 3738.25833927],
       [28190.41129687],
       [10600.67828764],
       [11395.63055992],
       [ 6248.83935385],
       [ 8423.19721152],
       [27360.73273479],
       [33843.62542848],
       [14432.50562897],
       [11759.04359894],
       [32070.2132407 ],
       [ 4152.83526684],
       [ 9364.57488075],
       [ 1227.5561631 ],
       [ 9712.31328139],
       [ 3714.56843696],
       [10585.88359557],
       [ 9162.47300707],
       [39963.69536536],
       [15580.1455371 ],
       [13795.75096539],
       [24960.11670118],
       [ 5274.13784945],
       [12507.41956494],
       [30840.51073591],
       [33738.39701967],
       [ 3693.97640602],


Evaluation Metrices are given below:

1. Mean Absolute Error (MAE)

2. Mean Squared Error (MSE)

3. R-Squared Error

In machine learning, absolute error refers to the magnitude of difference between the prediction of an observation and the true value of that observation. MAE takes the average of absolute errors for a group of predictions and observations as a measurement of the magnitude of errors for the entire group.

Mean squared error (MSE) is the average of sum of squared difference between actual value and the predicted or estimated value. It is also termed as mean squared deviation (MSD).

R-Squared is the ratio of Sum of Squares Regression (SSR) and Sum of Squares Total (SST). Sum of Squares Regression is amount of variance explained by the regression line. R-squared value is used to measure the goodness of fit. Greater the value of R-Squared, better is the regression model.

In [ ]:
from sklearn.metrics import mean_absolute_error

# MAE

mean_absolute_error(y_test, y_pred)

4012.715276853087

In [ ]:
from sklearn.metrics import mean_squared_error

mean_squared_error(y_test, y_pred)

32135747.239381798

In [ ]:
from sklearn.metrics import r2_score

r2_score(y_test, y_pred)

0.7958786376014413

So, r squared is 0.796 for multiple linear regression algorithm.

### Support Vector Regression

SVM or Support Vector Machine is a linear model for classification and regression problems. The idea of SVM is simple: The algorithm creates a line or a hyperplane which separates the data into classes.

In [ ]:
from sklearn.preprocessing import StandardScaler
X_sc = StandardScaler()
y_sc = StandardScaler()
X_train_new = X_sc.fit_transform(X_train[['age','bmi','children','smoker_no', 'smoker_yes','sex_female', 'sex_male','region_northeast', 'region_northwest','region_southeast', 'region_southwest']])
X_test_new = X_sc.transform(X_test[['age','bmi','children','smoker_no', 'smoker_yes','sex_female', 'sex_male','region_northeast', 'region_northwest','region_southeast', 'region_southwest']])
y_train_new = y_sc.fit_transform(y_train[['charges']])
y_test_new = y_sc.transform(y_test[['charges']])

In [ ]:

# Fitting SVR to the dataset
from sklearn.svm import SVR 
regressor = SVR(kernel = 'linear')
regressor.fit(X_train_new,y_train_new)

SVR(kernel='linear')

In [ ]:
y_pred = regressor.predict(X_test_new)

In [ ]:
mean_absolute_error(y_test_new, y_pred)

0.2722126001092648

In [ ]:
mean_squared_error(y_test_new, y_pred)

0.2708450958795668

In [ ]:
r2_score(y_test_new, y_pred)

0.7541715424584545

So, r squared is 0.754 for Support Vector Regression alogorithm.

### Decision Tree Regression
Decision tree builds regression or classification models in the form of a tree structure. It breaks down a dataset into smaller and smaller subsets while at the same time an associated decision tree is incrementally developed. The final result is a tree with decision nodes and leaf nodes.

In [ ]:
# Fitting the decision tree regression with the dataset
from sklearn.tree import DecisionTreeRegressor
regressor = DecisionTreeRegressor(random_state = 0)
regressor.fit(X_train,y_train) 


DecisionTreeRegressor(random_state=0)

In [ ]:
y_pred = regressor.predict(X_test)


In [ ]:
mean_absolute_error(y_test, y_pred)

3021.131019817911

In [ ]:
mean_squared_error(y_test, y_pred)

43846928.819191664

In [ ]:
r2_score(y_test, y_pred)

0.7214909993878144

So, r squared is 0.72 for Decision tree regression algorithm.

### Random Forest Regression
Random Forest Regression is a supervised learning algorithm that uses ensemble learning method for regression. A Random Forest operates by constructing several decision trees during training time and outputting the mean of the classes as the prediction of all the trees.

In [ ]:
# Fitting the Random Forest Regression with the dataset
from sklearn.ensemble import RandomForestRegressor

regressor = RandomForestRegressor(n_estimators = 10,random_state = 0) # n estiamator is the number of decision trees
regressor.fit(X_train,y_train)

RandomForestRegressor(n_estimators=10, random_state=0)

In [ ]:
y_pred = regressor.predict(X_test)

In [ ]:
mean_absolute_error(y_test, y_pred)

2763.557355874627

In [ ]:
mean_squared_error(y_test, y_pred)

21805383.963926636

In [ ]:
r2_score(y_test, y_pred)

0.8614955286651672

R squared = 0.796 (Multiple linear regression)

R squared = 0.754 (Support vector regression)

R squared = 0.72 (Decision tree regression)

R squared = 0.86 (Random forest Regression)

Thus, Random forest regression can predict the insurance more accurately rather than other regression algorithms. As greater the value of R-Squared, better is the regression model. So, it can be said that the performance of random forest regression is better than other algoritms for this dataset.